mtags/src/main/scala-3/scala/meta/internal/pc/InferredTypeProvider.scala

```diff
@@ -0,0 +1,261 @@
+package scala.meta.internal.pc
+
+import scala.meta.pc.OffsetParams
+import scala.meta.pc.PresentationCompilerConfig
+import scala.meta.tokens.{Token => T}
+
+import org.eclipse.lsp4j.TextEdit
+import java.nio.file.Paths
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.interactive.Interactive
+import java.net.URI
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.Spans
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.core.Types._
+import scala.meta.internal.mtags.MtagsEnrichments._
+import dotty.tools.dotc.core.Symbols.NoSymbol
+import scala.meta as m
+import scala.meta.internal.pc.AutoImports.AutoImportsGenerator
+import dotty.tools.dotc.ast.untpd
+
+/**
+ * Tries to calculate edits needed to insert the inferred type annotation
+ * in all the places that it is possible such as:
+ * - value or variable declaration
+ * - methods
+ * - pattern matches
+ * - for comprehensions
+ * - lambdas
+ *
+ * The provider will not check if the type does not exist, since there is no way to
+ * get that data from the presentation compiler. The actual check is being done via
+ * scalameta parser in InsertInferredType code action.
+ *
+ * @param params position and actual source
+ * @param driver Scala 3 interactive compiler driver
+ * @param config presentation compielr configuration
+ */
+final class InferredTypeProvider(
+    params: OffsetParams,
+    driver: InteractiveDriver,
+    config: PresentationCompilerConfig
+) {
+
+  def inferredTypeEdits(): List[TextEdit] = {
+    val uri = params.uri
+    val filePath = Paths.get(uri)
+    val source = SourceFile.virtual(filePath.toString, params.text)
+    driver.run(uri, source)
+    val unit = driver.currentCtx.run.units.head
```

Comment:
Same as the other comment, I'm assuming it's fine, but we don't need to worry about the `head` call here?

Reply:
We shouldn't need to worry, that head is the just parsed source file. We can add a check, but this is not the only place we are doing it and there seems to be no issues around it.

Reply:
No need to add a check then. I was just curious :)

Reply:
And this is basically produced by `driver.run(uri, source)`